### PR TITLE
RavenDB-14236: Upon occupying 25% of physical memory with scratches d…

### PR DESCRIPTION
**Pull request by aviviadi**
…irty memory, we start rejecting (http status code = 503, ServiceUnavailable) writes, until state is back below 25%. This solves the case where we have a machine with a lot of physical memory but really low write speed performing storage. The memory stores the scratches dirty and IO is not able to clear it fast enough.  This way we respond with 503, and let another node handle the requests while the IO struggles to write less data and return to normal operation.

### CODE REVIEW

File: src/Raven.Server/Documents/TransactionOperationsMerger.cs [hunk 1]
```diff
@@ -821,11 +821,31 @@ private enum PendingOperations
                     break;
 
                 pendingOps.Add(op);
-                meter.IncrementCounter(1);
 
+                var llt = context.Transaction.InnerTransaction.LowLevelTransaction;
+
+                var now = DateTime.UtcNow;
+                if (now - _lastHighDirtyMemCheck > TimeSpan.FromSeconds(10)) // we do not need to test scratch dirty mem every write
```
Comment (aviviadi) [inline src/Raven.Server/Documents/TransactionOperationsMerger.cs]: I thought about 10 seconds between checks while state is fine in order to have minimum affect on write speeds

File: src/Raven.Server/Config/Categories/MemoryConfiguration.cs [hunk 2]
```diff
@@ -45,5 +45,17 @@ public MemoryConfiguration()
...
+
+        [Description("The minimum amount of physical memory needed in order to operate 'High Dirty Memory' check mechanism. Default: 2048MB")]
+        [DefaultValue(2048)]
+        [SizeUnit(SizeUnit.Megabytes)]
+        [ConfigurationEntry("Memory.MinimumAllowedUseInMb", ConfigurationEntryScope.ServerWideOnly)]
+        public Size MinimumAllowedUseInMb { get; set; }
+
+        [Description("Specifies the threshold percentage of physical memory for activating 'High Dirty Memory' mechanism (server will return 'Service Unavailable' for writes when scratch files dirty memory exeeds this threshold). Default: 25")]
+        [DefaultValue(25)]
+        [SizeUnit(SizeUnit.Megabytes)]
```
Comment (ayende) [inline src/Raven.Server/Config/Categories/MemoryConfiguration.cs]: 25 what? MB, % ? 
The name of the property and its usage looks wrong

File: src/Raven.Server/Documents/TransactionOperationsMerger.cs (hunk 1 above)
Comment (ayende) [inline src/Raven.Server/Documents/TransactionOperationsMerger.cs]: Let's make the time configurable as well

Comment (ayende) [reply src/Raven.Server/Documents/TransactionOperationsMerger.cs]: And default to 30 seconds, I think.
This isn't something that we are likely to need high precision for

File: src/Raven.Server/Documents/TransactionOperationsMerger.cs [hunk 3]
```diff
@@ -821,11 +821,31 @@ private enum PendingOperations
...
+                    var minimumAllowedUseInBytes = _parent.Configuration.Memory.MinimumAllowedUseInMb.GetValue(SizeUnit.Bytes);
+                    var percentageFromPhysicalMem = _parent.Configuration.Memory.PercentageFromPhysicalMem;
+                    if (MemoryInformation.IsHighDirtyMemory(minimumAllowedUseInBytes, percentageFromPhysicalMem, out var details))
+                    {
+                        var highDirtyMemory = new HighDirtyMemoryException(
+                            $"Operation was cancelled by the transaction merger for transaction #{llt.Id} due to high dirty memory in scratch files." +
+                            $" This might be caused by a slow IO storage. Current memory usage: {details}");
+                        op.Exception = highDirtyMemory;
+                        var rejectedBuffer = GetBufferForPendingOps();
+                        rejectedBuffer.Add(op);
```
Comment (ayende) [inline src/Raven.Server/Documents/TransactionOperationsMerger.cs]: Here, need to reject _all_ the pending operations that we have, no?

Comment (ayende) [reply src/Raven.Server/Documents/TransactionOperationsMerger.cs]: `pendingOps`

File: src/Voron/Impl/Scratch/ScratchBufferFile.cs [hunk 4]
```diff
@@ -60,6 +65,10 @@ private void ClearDictionaries()
             _allocatedPages.Clear();
             _freePagesBySizeAvailableImmediately.Clear();
             _freePagesBySize.Clear();
+            if (MemoryInformation.MemoryMappedAllocatedMemory.TryRemove(_thisDebugInfo, out _) == false)
+            {
+                Console.WriteLine("ADIADI::Cannot remove " + _thisDebugInfo.DebugInfo + "," + _thisDebugInfo.scratchNumber);
```
Comment (ayende) [inline src/Voron/Impl/Scratch/ScratchBufferFile.cs]: Debug code

File: src/Voron/Impl/Scratch/ScratchBufferFile.cs [hunk 5]
```diff
@@ -43,6 +46,8 @@ public ScratchBufferFile(AbstractPager scratchPager, int scratchNumber)
             _scratchPager = scratchPager;
             _scratchNumber = scratchNumber;
             _allocatedPagesCount = 0;
+            _thisDebugInfo = (_scratchPager.DebugInfo, scratchNumber.ToString());
+            MemoryInformation.MemoryMappedAllocatedMemory[_thisDebugInfo] = _allocatedPagesCount * Constants.Storage.PageSize;
```
Comment (ayende) [inline src/Voron/Impl/Scratch/ScratchBufferFile.cs]: why do we need this?
Also, there may be multiple concurrent maps to the same file

File: src/Sparrow.Server/LowMemory/MemoryInformation.cs [hunk 6]
```diff
@@ -335,6 +343,15 @@ internal static MemoryInfoResult GetMemoryInfo(SmapsReader smapsReader = null, b
             }
         }
 
+        public static long GetTotalScratchAllocatedMemory()
```
Comment (ayende) [inline src/Sparrow.Server/LowMemory/MemoryInformation.cs:348]: This is wrong. You are computing the total size of the scratches, but you need the *dirty* size of them.
We already can tell that from the scrtaches themselves, so trying to do this roundabout way is actually more complex.
Have a way to find all the live scratches and ask them for their used mem

**Review by ayende — CHANGES_REQUESTED** [review]

File: src/Raven.Server/Config/Categories/MemoryConfiguration.cs [hunk 7]
```diff
@@ -45,5 +46,22 @@ public MemoryConfiguration()
...
+
+        [Description("The minimum amount of physical memory needed in order to operate 'High Dirty Memory' check mechanism. Default: 2048MB")]
+        [DefaultValue(2048)]
+        [SizeUnit(SizeUnit.Megabytes)]
+        [ConfigurationEntry("Memory.MinimumTemporaryDirtyMemoryUseAllowedInMb", ConfigurationEntryScope.ServerWideOnly)]
+        public Size MinimumTemporaryDirtyMemoryUseAllowedInMb { get; set; }
+
+        [Description("Specifies the threshold percentage of physical memory for activating 'High Dirty Memory' mechanism (server will return 'Service Unavailable' for writes when scratch files dirty memory exeeds this threshold). Default: 25")]
+        [DefaultValue(25)]
+        [ConfigurationEntry("Memory.TemporaryDirtyMemoryAllowedOutOfPhysicalMemoryInPercents", ConfigurationEntryScope.ServerWideOnly)]
```
Comment (ayende) [inline src/Raven.Server/Config/Categories/MemoryConfiguration.cs]:
```suggestion
        [ConfigurationEntry("Memory.PercentageOfTemporaryDirtyMemoryAllowed", ConfigurationEntryScope.ServerWideOnly)]
```

File: src/Raven.Server/Documents/TransactionOperationsMerger.cs [hunk 8]
```diff
@@ -821,11 +822,40 @@ private enum PendingOperations
                     break;
 
                 pendingOps.Add(op);
-                meter.IncrementCounter(1);
 
+                var llt = context.Transaction.InnerTransaction.LowLevelTransaction;
+
+                var now = DateTime.UtcNow;
+                if (now - _lastHighDirtyMemCheck > TimeSpan.FromSeconds(_parent.Configuration.Memory.TemporaryDirtyMemoryChecksPeriod)) // we do not need to test scratch dirty mem every write
```
Comment (ayende) [inline src/Raven.Server/Documents/TransactionOperationsMerger.cs]: Move the timespan creation to the ctor

File: src/Sparrow.Server/LowMemory/MemoryInformation.cs [hunk 9]
```diff
@@ -335,6 +343,18 @@ internal static MemoryInfoResult GetMemoryInfo(SmapsReader smapsReader = null, b
...
+        {
+            long totalScratchAllocated = 0;
+            foreach (var scratchGetAllocated in DirtyMemoryObjects)
+            {
+                if (scratchGetAllocated.Value != null)
```
Comment (ayende) [inline src/Sparrow.Server/LowMemory/MemoryInformation.cs]: Use: `totalScratchAllocated += scratchGetAllocated.Value?.Invoke()`

Right now, you have a race condition here that will kill us with NRE in dedicated thread

File: src/Sparrow.Server/LowMemory/MemoryInformation.cs [hunk 10]
```diff
@@ -636,6 +656,26 @@ private static void SetMemoryRecords(long availableRamInBytes)
             HighLastFiveMinutes = highLastFiveMinutes;
             LowLastFiveMinutes = lowLastFiveMinutes;
         }
+
+        public static bool IsHighDirtyMemory(long minimumAllowedUseInBytes, int percentageFromPhysicalMem, out string details)
```
Comment (ayende) [inline src/Sparrow.Server/LowMemory/MemoryInformation.cs]: Shouldn't the percentage be a double?

File: src/Sparrow.Server/LowMemory/MemoryInformation.cs [hunk 11]
```diff
@@ -636,6 +656,26 @@ private static void SetMemoryRecords(long availableRamInBytes)
...
+        {
+            details = null;
+            var totalScratchMemory = GetTotalScratchAllocatedMemory();
+            
+            if (totalScratchMemory <= (TotalPhysicalMemory.GetValue(SizeUnit.Bytes) * percentageFromPhysicalMem) / 100)
```
Comment (ayende) [inline src/Sparrow.Server/LowMemory/MemoryInformation.cs]: I don't like the separate checks.
We should error if _either_ percentage is too high or amount is too high.
Right now we error on amount only if percentage is too high

**Review by ayende — APPROVED** [review]

**Review by ayende — APPROVED** [review]

File: src/Raven.Server/Config/Categories/MemoryConfiguration.cs [hunk 12]
```diff
@@ -45,5 +49,21 @@ public MemoryConfiguration()
...
         public bool UseTotalDirtyMemInsteadOfMemUsage { get; set; }
+
+        [Description("EXPERT: Whether the high temporary dirty memory check is enabled. Default: true if the system has more than 2GB RAM")]
+        [DefaultValue(DefaultValueSetInConstructor)]
+        [ConfigurationEntry("Memory.EnableHighTemporaryDirtyMemoryUse", ConfigurationEntryScope.ServerWideOnly)]
+        public bool EnableHighTemporaryDirtyMemoryUse { get; set; }
+
+        [Description("Threshold percentage of memory for activating 'High Dirty Memory' mechanism (server will return 'Service Unavailable' for writes when scratch files dirty memory exeeds this threshold). Default: 25%")]
+        [DefaultValue(0.25d)]
+        [ConfigurationEntry("Memory.PercentageOfTemporaryDirtyMemoryAllowed", ConfigurationEntryScope.ServerWideOnly)]
```
Comment (ppekrol) [inline src/Raven.Server/Config/Categories/MemoryConfiguration.cs]: Let's be consistent with other configuration options. Rename to `TemporaryDirtyMemoryAllowedPercentage`

File: src/Raven.Server/Config/Categories/MemoryConfiguration.cs [hunk 13]
```diff
@@ -45,5 +49,21 @@ public MemoryConfiguration()
...
+        public double PercentageOfTemporaryDirtyMemoryAllowed { get; set; }
+
+        [Description("Period in seconds between 'High Dirty Memory' checks. Default: 30 Seconds")]
+        [DefaultValue(30)]
+        [TimeUnit(TimeUnit.Seconds)]
+        [ConfigurationEntry("Memory.TemporaryDirtyMemoryChecksPeriod", ConfigurationEntryScope.ServerWideOnly)]
```
Comment (ppekrol) [inline src/Raven.Server/Config/Categories/MemoryConfiguration.cs]: The config needs to have unit in the name: `TemporaryDirtyMemoryChecksPeriodInSec`
Property does not have to.

File: src/Raven.Server/Config/Categories/MemoryConfiguration.cs [hunk 14]
```diff
@@ -45,5 +49,21 @@ public MemoryConfiguration()
...
+        public int TemporaryDirtyMemoryChecksPeriod { get; set; }
```
Comment (ppekrol) [inline src/Raven.Server/Config/Categories/MemoryConfiguration.cs]: Should be TimeSetting

File: src/Raven.Server/Documents/TransactionOperationsMerger.cs [hunk 15]
```diff
@@ -60,6 +61,7 @@ public TransactionOperationsMerger(DocumentDatabase parent, CancellationToken sh
             _maxTxSizeInBytes = _parent.Configuration.TransactionMergerConfiguration.MaxTxSize.GetValue(SizeUnit.Bytes);
             _maxTimeToWaitForPreviousTxBeforeRejectingInMs = _parent.Configuration.TransactionMergerConfiguration.MaxTimeToWaitForPreviousTxBeforeRejecting.AsTimeSpan.TotalMilliseconds;
             _is32Bits = _parent.Configuration.Storage.ForceUsing32BitsPager || PlatformDetails.Is32Bits;
+            _timeToCheckHighDirtyMemory = TimeSpan.FromSeconds(_parent.Configuration.Memory.TemporaryDirtyMemoryChecksPeriod);
```
Comment (ppekrol) [inline src/Raven.Server/Documents/TransactionOperationsMerger.cs]: TemporaryDirtyMemoryChecksPeriod should be a TimeSetting

File: src/Raven.Server/Documents/TransactionOperationsMerger.cs [hunk 16]
```diff
@@ -821,11 +824,41 @@ private enum PendingOperations
                     break;
 
                 pendingOps.Add(op);
-                meter.IncrementCounter(1);
 
+                var llt = context.Transaction.InnerTransaction.LowLevelTransaction;
+
+                if (_parent.Configuration.Memory.EnableHighTemporaryDirtyMemoryUse)
+                {
+                    var now = DateTime.UtcNow;
```
Comment (ppekrol) [inline src/Raven.Server/Documents/TransactionOperationsMerger.cs]: _parent.Time.UtcNow

File: src/Raven.Server/Documents/TransactionOperationsMerger.cs [hunk 17]
```diff
@@ -1083,6 +1116,8 @@ public void Dispose()
         }
 
         private RecordingTx _recording = default;
+        private DateTime _lastHighDirtyMemCheck = DateTime.UtcNow;
```
Comment (ppekrol) [inline src/Raven.Server/Documents/TransactionOperationsMerger.cs]: _parent.Time.UtcNow

File: src/Raven.Server/Program.cs [hunk 18]
```diff
@@ -392,7 +392,7 @@ public static void WriteServerStatsAndWaitForEsc(RavenServer server)
         {
             var workingSetText = PlatformDetails.RunningOnPosix == false ? "working set" : "    RSS    ";
             Console.WriteLine("Showing stats, press any key to close...");
-            Console.WriteLine($"    {workingSetText}     | native mem      | managed mem     | mmap size         | reqs/sec       | docs (all dbs)");
+            Console.WriteLine($"    {workingSetText}     | native mem      | managed mem     | mmap size          | scratch dirty  | reqs/sec       | docs (all dbs)");
```
Comment (ppekrol) [inline src/Raven.Server/Program.cs:395]: Sanity check. Is this fine? There is an additional space between mmap size and scratch dirty and we are using positions for writing values here.

File: src/Sparrow.Server/LowMemory/MemoryInformation.cs [hunk 19]
```diff
@@ -315,16 +318,21 @@ internal static MemoryInfoResult GetMemoryInfo(SmapsReader smapsReader = null, b
...
+                    else if (PlatformDetails.RunningOnMacOsx)
+                        result = GetMemoryInfoMacOs(process, extended);
+                    else result = GetMemoryInfoLinux(smapsReader, extended);
+                }
 
-                    if (PlatformDetails.RunningOnMacOsx)
-                        return GetMemoryInfoMacOs(process, extended);
+                var totalScratchAllocated = GetTotalScratchAllocatedMemory();
+
+                result.AvailableMemory -= new Size(totalScratchAllocated, SizeUnit.Bytes);
```
Comment (ppekrol) [inline src/Sparrow.Server/LowMemory/MemoryInformation.cs]: FYI. You can do: `result.AvailableMemory.Add(-totalScratchAllocated, SizeUnit.Bytes)`

File: src/Raven.Server/Program.cs (hunk 18 above)
Comment (aviviadi) [reply src/Raven.Server/Program.cs:395]: This space is needed

**Review by ppekrol — APPROVED** [review]

**Pull request merged**